packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,500 @@
+      // Nothing to do here for EVM accounts
+      return;
+    }
+    const releaseLock = await this.#mutex.acquire();
+
+    try {
+      // Get assets list
+      if (account.metadata.snap) {
+        const assets = await this.#getAssetsList(
+          account.id,
+          account.metadata.snap.id,
+        );
+        await this.#refreshAssetsMetadata(assets);
+        this.update((state) => {
+          state.allNonEvmTokens[account.id] = assets;
+        });
+      }
+    } finally {
+      releaseLock();
+    }
+  }
+
+  /**
+   * Handles changes when a new account has been removed.
+   *
+   * @param accountId - The new account id being removed.
+   */
+  async #handleOnAccountRemoved(accountId: string): Promise<void> {
+    // Check if accountId is in allNonEvmTokens and if it is, remove it
+    if (this.state.allNonEvmTokens[accountId]) {
+      this.update((state) => {
+        delete state.allNonEvmTokens[accountId];
+      });
+    }
+  }
+
+  /**
+   * Refreshes the assets snaps and metadata for the given list of assets
+   *
+   * @param assets - The assets to refresh
+   */
+  async #refreshAssetsMetadata(assets: CaipAssetType[]) {
+    const assetsWithoutMetadata: CaipAssetType[] = assets.filter(
+      (asset) => !this.state.metadata[asset],
+    );
+
+    // call the snap to get the metadata
+    if (assetsWithoutMetadata.length > 0) {
+      // check if for every asset in assetsWithoutMetadata there is a snap in snaps by chainId else call getAssetSnaps
+      if (
+        !assetsWithoutMetadata.every((asset: CaipAssetType) => {
+          const { chainId } = parseCaipAssetType(asset);
+          return Boolean(this.#getAssetSnapFor(chainId));
+        })
+      ) {
+        this.#snaps = this.#getAssetSnaps();
+      }
+      await this.#updateAssetsMetadata(assetsWithoutMetadata);
+    }
+  }
+
+  /**
+   * Updates the assets metadata for the given list of assets
+   *
+   * @param assets - The assets to update
+   */
+  async #updateAssetsMetadata(assets: CaipAssetType[]) {
+    // Creates a mapping of scope to their respective assets list.
+    const assetsByScope: Record<CaipChainId, CaipAssetType[]> = {};
+    for (const asset of assets) {
+      const { chainId } = parseCaipAssetType(asset);
+      if (!assetsByScope[chainId]) {
+        assetsByScope[chainId] = [];
+      }
+      assetsByScope[chainId].push(asset);
+    }
+    let newMetadata: Record<CaipAssetType, FungibleAssetMetadata> = {};
+
+    for (const chainId of Object.keys(assetsByScope)) {
+      const assetsForChain = assetsByScope[chainId as CaipChainId];
+      // Now fetch metadata from the associated asset Snaps:
+      const snap = this.#getAssetSnapFor(chainId as CaipChainId);
+      if (snap) {
+        const metadata = await this.#getAssetsMetadataFrom(
+          assetsForChain,
+          snap.id,
+        );
+        newMetadata = {
+          ...newMetadata,
+          ...(metadata?.assets ?? {}),
+        };
+      }
+    }
```

Comment:
Type-cast should not be needed?

```suggestion

    let newMetadata: Record<CaipAssetType, FungibleAssetMetadata> = {};
    for (const assetsForChain of Object.keys(assetsByScope)) {
      const assetsForChain = assetsByScope[chainId];
      // Now fetch metadata from the associated asset Snap:
      const snap = this.#getAssetSnapFor(chainId);
      if (snap) {
        const metadata = await this.#getAssetsMetadataFrom(
          assetsForChain,
          snap.id,
        );
        newMetadata = {
          ...newMetadata,
          ...(metadata?.assets ?? {}),
        };
      }
    }
```

Reply:
yeah that one is related to this comment https://github.com/MetaMask/core/pull/5138#discussion_r1939714908 still not sure why its not being inferred correctly

Reply:
Updated here [33c1352](https://github.com/MetaMask/core/pull/5138/commits/33c13522d0adacb5ea22f9b34cdfd0b5772d40c0) 🙏